diff --git a/host/gl/ColorBufferGl.cpp b/host/gl/ColorBufferGl.cpp
index d0129b2..2bd56c2 100644
--- a/host/gl/ColorBufferGl.cpp
+++ b/host/gl/ColorBufferGl.cpp
@@ -833,6 +833,8 @@ bool ColorBufferGl::bindToTexture() {
         return false;
     }
 
+    waitSync();
+
     if (tInfo->currContext->clientVersion() > GLESApi_CM) {
         s_gles2.glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_eglImage);
     } else {
@@ -846,6 +848,8 @@ bool ColorBufferGl::bindToTexture2() {
         return false;
     }
 
+    waitSync();
+
     s_gles2.glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_eglImage);
     return true;
 }
diff --git a/host/gl/glestranslator/GLES_V2/GLESv30Imp.cpp b/host/gl/glestranslator/GLES_V2/GLESv30Imp.cpp
index fea7f60e..2d603a69 100644
--- a/host/gl/glestranslator/GLES_V2/GLESv30Imp.cpp
+++ b/host/gl/glestranslator/GLES_V2/GLESv30Imp.cpp
@@ -983,36 +983,7 @@ GL_APICALL void GL_APIENTRY glProgramBinary(GLuint program, GLenum binaryFormat,
     GET_CTX_V2();
     if (ctx->shareGroup().get()) {
         const GLuint globalProgramName = ctx->shareGroup()->getGlobalName(NamedObjectType::SHADER_OR_PROGRAM, program);
-        SET_ERROR_IF(globalProgramName == 0, GL_INVALID_VALUE);
-
-        auto objData =
-            ctx->shareGroup()->getObjectData(NamedObjectType::SHADER_OR_PROGRAM, program);
-        SET_ERROR_IF(!objData, GL_INVALID_OPERATION);
-        SET_ERROR_IF(objData->getDataType() != PROGRAM_DATA, GL_INVALID_OPERATION);
-
-        ProgramData* programData = (ProgramData*)objData;
-
         ctx->dispatcher().glProgramBinary(globalProgramName, binaryFormat, binary, length);
-
-        GLint linkStatus = GL_FALSE;
-        ctx->dispatcher().glGetProgramiv(globalProgramName, GL_LINK_STATUS, &linkStatus);
-
-        programData->setHostLinkStatus(linkStatus);
-        programData->setLinkStatus(linkStatus);
-
-        GLsizei infoLogLength = 0;
-        ctx->dispatcher().glGetProgramiv(globalProgramName, GL_INFO_LOG_LENGTH, &infoLogLength);
-
-        if (infoLogLength > 0) {
-            std::vector<GLchar> infoLog(infoLogLength);
-            ctx->dispatcher().glGetProgramInfoLog(globalProgramName, infoLogLength, &infoLogLength,
-                                                  infoLog.data());
-
-            if (infoLogLength) {
-                infoLog.resize(infoLogLength);
-                programData->setInfoLog(infoLog.data());
-            }
-        }
     }
 }
 
